internal/job/shell/tempfile.go

```diff
@@ -1,50 +0,0 @@
-package shell
-
-import (
-	"fmt"
-	"os"
-	"path/filepath"
-	"strings"
-)
-
-// TempFileWithExtension creates a temporary file that copies the extension of the provided filename
-func TempFileWithExtension(filename string) (*os.File, error) {
-	extension := filepath.Ext(filename)
-	basename := strings.TrimSuffix(filename, extension)
-
-	// TempDir is not guaranteed to exist
-	tempDir := os.TempDir()
-	if _, err := os.Stat(tempDir); os.IsNotExist(err) {
-		// Actual file permissions will be reduced by umask, and won't be 0777 unless the user has manually changed the umask to 000
-		if err = os.MkdirAll(tempDir, 0777); err != nil {
-			return nil, err
-		}
-	}
-
-	// Create the file
-	tempFile, err := os.CreateTemp(tempDir, basename+"-")
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create temporary file \"%s\" (%s)", filename, err)
-	}
-
-	// Do we need to rename the file?
-	if extension != "" {
-		// Close the currently open tempfile
-		tempFile.Close()
-
-		// Rename it
-		newTempFileName := tempFile.Name() + extension
-		err = os.Rename(tempFile.Name(), newTempFileName)
-		if err != nil {
-			return nil, fmt.Errorf("Failed to rename \"%s\" to \"%s\" (%s)", tempFile.Name(), newTempFileName, err)
-		}
-
-		// Open it again
-		tempFile, err = os.OpenFile(newTempFileName, os.O_RDWR|os.O_EXCL, 0600)
```

Comment:
It seems strange to use `O_EXCL` here, as the file was created by the earlier call to `os.CreateTemp` and then (maybe) renamed to its current location. So the file should already exist, and this may be expected to error.

However, turns [out](https://man7.org/linux/man-pages/man2/open.2.html#DESCRIPTION):
> In general, the behavior of O_EXCL is undefined if it is used without O_CREAT. There is one exception: on Linux 2.6 and later, O_EXCL can be used without O_CREAT if pathname refers to a block device. If the block device is in use by the system (e.g., mounted), open() fails with the error EBUSY.

So this may have had no effect.

Reply:
We avoid all this in the replacement code in `internal/tmpfile/tmpfile.go`.